data/tutorials/language/1ms_01_functors.md

```diff
@@ -8,183 +8,381 @@ category: "Module System"
-the future, which is much worse.
+**Prerequisites**: Transitive closure leading to modules.
 
-## Using an Existing Functor
+## Project Setup
 
-The standard library defines a `Set` module, which provides a `Make` functor.
-This functor takes one argument, which is a module that provides (at least) two
-things: the type of elements, given as `t` and the comparison function given as
-`compare`. The point of the functor is to ensure that the same comparison
-function will always be used, even if the programmer makes a mistake.
+This tutorial uses the [Dune](https://dune.build) build tool. Make sure to have installed version 3.7 or later. We start by creating a fresh project. We need a folder named `funkt` with files `dune-project`, `dune`, `funkt.opam` and `funkt.ml`, the latter two are created empty.
+```shell
+$ mkdir funkt; cd funkt
 
-For example, if we want to use sets of `ints`, we would do this:
+$ touch funkt.opam funkt.ml
+```
+
+**`dune-project`**
+```lisp
+(lang dune 3.7)
+```
+
+**`dune`**
+```lisp
+(executable
+  (name funkt)
+  (public_name funkt)
+  (libraries str))
+```
+
+Check this works using the `dune exec funkt` command, it shouldn't do anything (the empty file is valid OCaml syntax) but it shouldn't fail either. The stanza `libraries str` will be used later.
+
+## Using an Existing Functor: `Set.Make`
+
+The standard library contains a [`Set`](/api/Set.html) module providing a data structure that allows operations like union and intersection. To use the provided type and its associated [functions](/api/Set.S.html), it is required to use the functor provided by `Set`. For reference only, here is a shortened version of the interface of `Set`.
+```ocaml
+module type OrderedType = sig
+  type t
+  val compare : t -> t -> int
+end
+
+module type S = sig
+  (* ... *)
+end
+
+module Make : functor (Ord : OrderedType) -> S
+```
+
+The functor `Set.Make` needs to be passed a module of type `Set.OrderedType` to produce a module of type `Set.S`. Here is how it can look like in our project:
+
+**`funkt.ml`**
+```ocaml
+module StringSet = Set.Make(String)
+```
+
+With this, the command `dune exec funkt` shouldn't do anything but it shouldn't fail either. Here is the meaning of that statement
+- The module `funkt.StringSet` is defined
+- The module `String` (from the standard library) is applied to the functor `Set.Make`. This is allowed because the `String` module satisfies the interface `Set.OrderedSet`
+  - It defines a type name `t` (which is an alias for `string`)
+  - It defines a function `compare` of type `t -> t -> bool`, that is the function `String.compare`
+- The result module from the functor application `Set.Make(String)` is bound to the name `StringSet`, it has the signature `Set.S`.
+
+Add some code to the `funkt.ml` file to produce an executable that does something and check the result.
+
+**`funkt.ml`**
+```ocaml
+module StringSet = Set.Make(String)
+
+let _ =
```

Comment:
I think the whole parsing inputs and splitting by regexp is not adding to working with our functor here, and a hardcoded example would do just the same.

```ocaml
let () = 
  let set = StringSet.of_list ["hello", "world", "world"] in
  StringSet.iter (fun line -> print_endline line) set
```

BUT this makes me think that we haven't given functors a good raison d'être – like why do we even have functors to begin with?

Why does `Set` must be a functor but `Hashtbl` doesn't need to? 

And why does `Hashtbl` offer a `Hashtbl.Make` functor?

I'm realizing I don't know exactly what this is but let's say its because:
* we don't have a polymorphic `compare`
* we do have a polymorphic `hash`

If we had a polymorphic `compare` would we be using `Set` as an example here?

Reply:
This will be answered afterwards in the “parametrized implementation”. However, we can add a few words saying that being able to use a custom comparison function in all the functions of some sort of set, at once, is justification for set-as-functor.